[13.0] Add stock_storage_type_buffer

guewen opened the pull request:
This is an extensions for putaway with storage types (provided by
``stock_storage_type``).

In a warehouse, some areas where goods are put cannot be accessed
directly and need to be moved to a buffer first (for instance, because
they need a special machine to move from the buffer to the bins).

This module allows to define buffers in front of a set of locations.
When a buffer contains something or has move lines that reaches it, all
the locations are removed from the put-away computation.

Commit:
Add stock_storage_type_buffer

This is an extensions for putaway with storage types (provided by
``stock_storage_type``).

In a warehouse, some areas where goods are put cannot be accessed
directly and need to be moved to a buffer first (for instance, because
they need a special machine to move from the buffer to the bins).

This module allows to define buffers in front of a set of locations.
When a buffer contains something or has move lines that reaches it, all
the locations are removed from the put-away computation.

Commit:
Extract methods in stock_storage_type to allow extensions

simahawk commented on stock_storage_type_buffer/models/stock_location_storage_buffer.py:
```diff
@@ -0,0 +1,113 @@
+        leaf_locations = self.location_ids.leaf_location_ids
+        leaf_location_count = len(leaf_locations)
+
+        message = "<p>The buffer locations:<ul><br/>"
```
would be better to use a qweb template

simahawk commented on stock_storage_type_buffer/models/stock_location_storage_buffer.py:
```diff
@@ -0,0 +1,113 @@
+        result = []
+        for record in self:
+            name = ", ".join(
+                self.buffer_location_ids[: self.MAX_LOCATIONS_IN_NAME_GET].mapped(
```
```suggestion
                rec.buffer_location_ids[: self.MAX_LOCATIONS_IN_NAME_GET].mapped(
```

simahawk commented on stock_storage_type_buffer/models/stock_location_storage_buffer.py:
```diff
@@ -0,0 +1,113 @@
+                    "name"
+                )
+            )
+            if len(self.buffer_location_ids) > self.MAX_LOCATIONS_IN_NAME_GET:
```
```suggestion
            if len(rec.buffer_location_ids) > self.MAX_LOCATIONS_IN_NAME_GET:
```

simahawk commented on stock_storage_type/models/stock_location.py:
```diff
@@ -312,9 +312,39 @@ def _domain_location_storage_type_constraints(
+            valid_location_ids |= set(locations.ids)
+        return self.browse(valid_location_ids)
+
+    def _select_valid_locations(self, limit=None):
```
"select valid locations"... for what? I think is not clear here. Maybe `_select_valid_putaway_locations`?

guewen commented:
@simahawk could you review my fixups, then I squash them?

simahawk commented on stock_storage_type_buffer/models/stock_location_storage_buffer.py:
```diff
@@ -0,0 +1,90 @@
+    def _help_message_location_items(self, records):
+        items = records[: self.MAX_LOCATIONS_IN_HELP].mapped("display_name")
+        if len(records) > self.MAX_LOCATIONS_IN_HELP:
+            items.append("... and {} other locations.".format(
```
make it translatable or move this to the template as well using a `remaining_[locations|buffer_locations]_count` key in rendering values.

guewen replied:
I feel it would imply too much logic in the template with the qweb magic keys, so I ended up with providing the "remaining count" to the template, which seems simple and clean actually. Thanks for pointing this

guewen replied:
@simahawk and I pushed a last fixup that use a sub-template to avoid repeating "... and {} other locations." in the 2 places

simahawk commented on stock_storage_type_buffer/templates/stock_location_storage_buffer_templates.xml:
```diff
@@ -0,0 +1,49 @@
+                    </li>
+                </t>
+                <t t-if="buffer_locations_remaining">
+                    <t t-set="remaining" t-value="buffer_locations_remaining" />
```
or
```xml
<t t-call="stock_storage_type_buffer.storage_buffer_help_message_other_locations">
    <t t-set="remaining" t-value="buffer_locations_remaining" />
</t>
```
not blocking, of course

guewen replied:
Oh nice! so much better indeed, fixed and squashed

jgrandguillaume reviewed (approved):
This is just so elegant! Thanks for the great job here.

jbaudoux commented on stock_storage_type_buffer/models/stock_location_storage_buffer.py:
```diff
@@ -0,0 +1,88 @@
+    def buffers_have_capacity(self):
+        self.ensure_one()
+        buffer_locations = self.buffer_location_ids.leaf_location_ids
+        return any(location.location_is_empty for location in buffer_locations)
```
The buffer should also be considered empty as soon as
sum(location.out_move_lines_ids.qty_done) = sum(location.quant_ids.mapped("quantity"))
Probably to put in _compute_location_is_empty.
So when someone takes what is on the buffer by writing the done quantity on the move line, the buffer is physically empty and we should be able to put again something on it.
This detail is quite important for the selection of the right set of allowed locations.

guewen replied:
@jbaudoux can you check if this is what you expect: https://github.com/OCA/wms/pull/58/commits/aa7473a1f464a5a1e6078df79a8406510a1658c1 ?

jbaudoux replied:
Perfect. Thanks

jbaudoux commented on stock_storage_type_buffer/models/stock_location.py:
```diff
@@ -0,0 +1,54 @@
+
+    def _filter_locations_blocked_by_buffer(self):
+        valid_location_ids = set(self.ids)
+        for storage_buffer in self.env["stock.location.storage.buffer"].search([]):
```
You compute all the buffers availability even if the valid locations are not guarded by any buffer. I would have done it the other way around by checking each valid location if it is guarded by a buffer and filter it out if the buffer is not empty.

The pro:
- You compute only the required buffers. For each buffer location_is_empty should only be computed once if kept in cache.
- If you define the buffer on the location, you can read_group it at once to not read each location one by one which would be costly
- If there is a limit, you can stop the iteration on the locations when the limit is reached

guewen replied:
We have to measure the cost of the buffer availability, we don't have a lot of them and it only queries indexed relations.

guewen replied:
The cost of `if storage_buffer.buffers_have_capacity():` is really low (between 0.01 and 0.06 per buffer on my test database after stress test), so for now I don't think it's worth changing this.

Commit:
Add small optimization on sorted intersection

Commit:
Add outgoing move lines qty_done qty in "is_empty"

A location is considered empty if all the quantity of its quants
are already moved by the "qty_done" of the outgoing move lines.

simahawk commented:
@jbaudoux ping :) @guewen any test to be added to make coverage happier?

simahawk commented:
@guewen ping

guewen commented:
The missing test coverage is on the help message and `name_get`. Is it worth to test these?

Commit:
Add tests for storage buffer name/help message

guewen commented:
Added tests and rebased

simahawk commented:
/ocabot merge nobump
